host.py

```diff
@@ -0,0 +1,87 @@
+        # maybe it should have a list of packets to send, a list of packets
+        # recieved and not processed yet, and a set of packets sent and still
+        # waiting for acks
+        self.packets_to_send = []
+        # map from packet id to packet
+        # TODO(agf): This relies on packet ids being unique over all the
+        # packets sent by any one host
+        self.packets_waiting_for_acks = dict()
+        self.next_unused_packet_id = 0
+        self.in_sending_chain = False
+
+    def generate_packet(self, dst_id):
+        self.packets_to_send.append(Packet(id_=self.next_unused_packet_id,
+                                           src=self.id_,
+                                           dst=dst_id,
+                                           size=globals_.DATA_PACKET_SIZE,
+                                           ack=False))
+        self.next_unused_packet_id += 1
+
+    def generate_packets_to_send(self, dst_id, how_much_data):
+        npackets = how_much_data // globals_.DATA_PACKET_SIZE
+        if how_much_data % globals_.DATA_PACKET_SIZE != 0:
+            npackets += 1
+        for _ in xrange(npackets):
+            self.generate_packet(dst_id)
+
+    def receive_packet(self, packet):
+        assert packet.dst == self.id_
+        if packet.ack:
+            globals_.event_manager.log('{} received ack for {}'.format(self.id_, packet))
+            assert packet.id_ in self.packets_waiting_for_acks
+            original_packet = self.packets_waiting_for_acks[packet.id_]
+            if packet.src == original_packet.dst and packet.dst == original_packet.src:
+                del self.packets_waiting_for_acks[packet.id_]
+        else:
+            # Send an ack
+            ack = Packet(id_=packet.id_,
+                         src=self.id_,
+                         dst=packet.src,
+                         size=globals_.ACK_SIZE,
+                         ack=True)
+            # Schedule the ack to be the next packet sent
+            self.packets_to_send.insert(0, ack)
+
+    def send_packet(self, packet, port):
+        self.ports[port].receive_from_device(packet)
+        if not packet.ack:
+            self.packets_waiting_for_acks[packet.id_] = packet
+            def ack_is_due():
+                if packet.id_ in self.packets_waiting_for_acks:
+                    globals_.event_manager.log(
+                            '{} is due to receive an ack for {}, has not received it'.format(
+                                self.id_, packet.id_))
+                    self.packets_to_send.insert(0, packet)
+                else:
+                    globals_.event_manager.log(
+                            '{} is due to receive an ack for {}, has already received it'.format(
+                                self.id_, packet.id_))
+            globals_.event_manager.add(globals_.TIME_TO_WAIT_UNTIL_ACK, ack_is_due)
+
+    def act(self):
+        if not self.in_sending_chain and self.packets_to_send:
```

Comment:
This `self.in_sending_chain` state seems weird, but I haven't thought of another way that I'm happy with. In the current approach, each send event (when executed) recursively schedules another send event to occur at time `globals_.TIME_BETWEEN_SENDS` in the future. We need to keep trach of `self.in_sending_chain` so that extra act() calls don't cause extra sends.

An alternative would be to keep track of the last time that we sent something and have the the act() function send a packet if it has been long enough. To make sure that we have a chance to act() when it's been just long enough, we would schedule a no-op at time `globals_.TIME_BETWEEN_SENDS` in the future, after each send. This would be cleaner, but it would break when the floating-point arithmetic is not exact.